services/brig/src/Brig/User/API/Auth.hs

```diff
@@ -185,11 +191,24 @@ ssoLogin (req ::: persist ::: _) = do
     a <- Auth.ssoLogin l typ !>> loginError
     tokenResponse a
 
-logout :: JSON ::: Maybe ZAuth.UserToken ::: Maybe ZAuth.AccessToken -> Handler Response
+legalHoldLogin :: JsonRequest LegalHoldLogin ::: JSON -> Handler Response
+legalHoldLogin (req ::: _) = do
+    l <- parseJsonBody req
+    let typ = PersistentCookie -- Session cookie isn't a supported use case here
+    a <- Auth.legalHoldLogin l typ !>> legalHoldLoginError
+    tokenResponse a
+
+-- TODO: add legalhold test checking cookies are revoked (/access/logout is called) when legalhold device is deleted.
+logout :: JSON ::: Maybe (Either ZAuth.UserToken ZAuth.LegalHoldUserToken) ::: Maybe (Either ZAuth.AccessToken ZAuth.LegalHoldAccessToken) -> Handler Response
 logout (_ ::: Nothing ::: Nothing) = throwStd authMissingCookieAndToken
 logout (_ ::: Nothing ::: Just _ ) = throwStd authMissingCookie
 logout (_ ::: Just _  ::: Nothing) = throwStd authMissingToken
-logout (_ ::: Just ut ::: Just at) = do
+logout (_ ::: Just (Left _) ::: Just (Right _)) = throwStd authTokenMismatch
+logout (_ ::: Just (Right _) ::: Just (Left _)) = throwStd authTokenMismatch
+logout (_ ::: Just (Left ut) ::: Just (Left at)) = do
+    Auth.logout ut at !>> zauthError
+    return empty
+logout (_ ::: Just (Right ut) ::: Just (Right at)) = do
```

Comment:
```suggestion
logout (_ ::: Just (either id id -> ut) ::: Just (either id id -> at)) = do
```

And then you could save yourself writing the morally redundant `Left` case.  But arguably your code is way more readable.

Reply:
nope, sorry, that wouldn't type-check.  i think there may be a solution down that rabbit-hole, but on second thoughts i'm against it.  :)